server/core/src/main/java/dev/slimevr/osc/VRCOSCHandler.kt

```diff
@@ -0,0 +1,485 @@
+		val currentTime = System.currentTimeMillis().toFloat()
+
+		// Gets timer from vrServer
+		if (fpsTimer == null) {
+			fpsTimer = VRServer.instance.fpsTimer
+		}
+		// Update received trackers' offset rotation slerp
+		if (receivingRotationOffset != receivingRotationOffsetGoal) {
+			receivingRotationOffset = receivingRotationOffset.interpR(receivingRotationOffsetGoal, OFFSET_SLERP_FACTOR * (fpsTimer?.timePerFrame ?: 1f))
+		}
+
+		// Send OSC data
+		if (oscSender != null && oscSender!!.isConnected) {
+			// Create new bundle
+			val bundle = OSCBundle()
+
+			for (i in computedTrackers.indices) {
+				if (trackersEnabled[i]) {
+					// Send regular trackers' positions
+					val (x, y, z) = computedTrackers[i].position
+					oscArgs.clear()
+					oscArgs.add(x)
+					oscArgs.add(y)
+					oscArgs.add(-z)
+					bundle.addPacket(
+						OSCMessage(
+							"/tracking/trackers/${getVRCOSCTrackersId(computedTrackers[i].trackerPosition)}/position",
+							oscArgs.clone()
+						)
+					)
+
+					// Send regular trackers' rotations
+					val (w, x1, y1, z1) = computedTrackers[i].getRotation()
+					// We flip the X and Y components of the quaternion because
+					// we flip the z direction when communicating from
+					// our right-handed API to VRChat's left-handed API.
+					// X quaternion represents a rotation from y to z
+					// Y quaternion represents a rotation from z to x
+					// When we negate the z direction, X and Y quaternion
+					// components must be negated.
+					val (_, x2, y2, z2) = Quaternion(
+						w,
+						-x1,
+						-y1,
+						z1
+					).toEulerAngles(EulerOrder.YXZ)
+					oscArgs.clear()
+					oscArgs.add(x2 * FastMath.RAD_TO_DEG)
+					oscArgs.add(y2 * FastMath.RAD_TO_DEG)
+					oscArgs.add(z2 * FastMath.RAD_TO_DEG)
+					bundle.addPacket(
+						OSCMessage(
+							"/tracking/trackers/${getVRCOSCTrackersId(computedTrackers[i].trackerPosition)}/rotation",
+							oscArgs.clone()
+						)
+					)
+				}
+				if (computedTrackers[i].trackerPosition === TrackerPosition.HEAD) {
+					// Send HMD position
+					val (x, y, z) = computedTrackers[i].position
+					oscArgs.clear()
+					oscArgs.add(x)
+					oscArgs.add(y)
+					oscArgs.add(-z)
+					bundle.addPacket(
+						OSCMessage(
+							"/tracking/trackers/head/position",
+							oscArgs.clone()
+						)
+					)
+				}
+			}
+
+			try {
+				oscSender!!.send(bundle)
+			} catch (e: IOException) {
+				// Avoid spamming AsynchronousCloseException too many
+				// times per second
+				if (currentTime - timeAtLastError > 100) {
+					timeAtLastError = System.currentTimeMillis()
+					LogManager
+						.warning(
+							"[VRCOSCHandler] Error sending OSC message to VRChat: " +
+								e
+						)
+				}
+			} catch (e: OSCSerializeException) {
+				if (currentTime - timeAtLastError > 100) {
+					timeAtLastError = System.currentTimeMillis()
+					LogManager
+						.warning(
+							"[VRCOSCHandler] Error sending OSC message to VRChat: " +
+								e
```

Comment:
I see the code was already like this before, but can't `e` be passed as an argument? It should be using string interpolation (i.e. `"error: $e"`) otherwise anyways.

Reply:
Done